GCRCatalogs/register.py

```diff
@@ -33,6 +49,16 @@ def load_yaml(yaml_file):
         config = yaml.safe_load(r.raw)
     return config
 
+def _resolve_dict(d):
+    for (k,v) in d.items():
+        if k in _PATH_LIKE:
+            d[k] = os.path.join(str(_ROOT_DIR), str(v))
+        elif k == 'catalogs':
+            # list of items, each of which is catalog spec.
+            for c in v:
+                _resolve_dict(c)
+    return d
```

Comment:
```suggestion
def _resolve_dict(d):
    for k, v in d.items():
        if isinstance(v, dict):
            d[k] = _resolve_dict(v)
        elif k in _PATH_LIKE and is_string_like(v):
            d[k] = os.path.join(ROOT_DIR, v)
    return d
```

Reply:
A couple comments:
1. I don't think this will handle the 'catalogs' case properly.  The structure of 'catalogs' is a list of dicts.  In a later version, not yet pushed, I made it a bit more general with a declaration 
`_DICT_LIST = ('catalogs')`
More generally, I'm somewhat torn about how general to make this code.  I considered just passing all lists and dicts through the code, but some of the config files are quite long, containing things which really don't need to be parsed, so I've compromised with tuples of known keywords of interest.
2. Originally I didn't cast _ROOT_DIR and v to str and one of the checks failed when I pushed to GitHub.  is_string_like seems obscure (the only place I found it so far is https://networkx.github.io/documentation/stable/reference/generated/networkx.utils.misc.is_string_like.html ).  I would rather check by using calls to isinstance.

Reply:
Sorry, `is_string_like` is already in `GCRCatalogs.utils`. 

You're right about the list issue. Let me think about it.

Reply:
How about this:

```python
def _prepend_roor_dir(value, key=None):
    if isinstance(value, dict):
        return {k: _prepend_roor_dir(v, k) for k, v in value.items()}
        
    if isinstance(value, list):
        return [_prepend_roor_dir(v, key) for v in value]
    
    if key in _PATH_LIKE and is_string_like(value):
       return os.path.join(ROOT_DIR, value)
   
    return value
```

Reply:
As I said - yes, you can do something like this.  But it means parsing a lot more of the content of config files than will ever be necessary.  Maybe the parsing is fast and it doesn't matter.

Reply:
The parsing has been done by yaml.safe_load anyway. Walking the dict is not costly. I think the `elif k == 'catalogs'` is a bit opaque as people may not know why it needs to be a special case. If we are going that route we need to add a bit more comment.

Reply:
My current draft  has
`_DICT_LIST = ('catalogs')`
and then a clause
```
        elif k in _DICT_LIST:
            # list of dicts, each of which may have _PATH_LIKE keys
            for c in v:
                _resolve_dict(c)
```
In choosing between this kind of thing and a more general handler like the one you propose there are two things to consider:

- is there a noticeable performance penalty?  The initial parsing is done by yaml.safe_load. GCR code would still have to look at every keyword with this approach, but maybe that's not a problem
- should we keep a tighter rein on the form of the config files?  
For example, someone could add a new keyword with a dict as value and one of the dict keys could happen to be in the _PATH_LIKE keys, but maybe not intended to be handled as those keys are at the top level or with a `catalogs` block.  

One way to avoid confusion would be to require $ROOT_DIR before any value which should be transformed.  If we go with the general handler, I think we need an explicit signal of some kind.

Reply:
On the other point about `is_string_like`  - since it's in the package it's reasonable to use it, but I suspect the check will still complain about  `os.path.join` line unless I cast the value, whereas if I use `isinstance(value, str)` it probably won't.   What is the advantage of `is_string_like`?

Reply:
`is_string_like` is mostly for Python 2-3 compatibility, so it's not as relevant now TBH. I am using it merely because there's already a function in `utils`. 

It's ok to turn off some linter checks by comments.

Reply:
On a different note, I suggest that we use `_PATH_LIKE_KEYS` or `_PATH_DEFINING_KEYS` instead just `_PATH_LIKE`.

Reply:
The reason I don't quite like the proposed solution of `_DICT_LIST = ('catalogs',)` is that I find it quite obscure --- I think it is difficult for new devs to figure out why this particular keyword `catalogs` is in `_DICT_LIST`. The reason that I proposed a more generic solution is that it should brings up fewer questions IMO. It is also not clear what happen in a new reader use `catalogs` to specify file path.  

The keywords in config files are not controlled or checked (with the exception of `subclass_name`, `alias`, and `based_on`) in `register.py`. They are just passed to each reader. Hence, the types that keywords corresponds to are determined by the readers, not by `register.py`. That's why this `_DICT_LIST` check makes me uncomfortable, as it creates more cross-dependencies between `register.py` and readers. 

If we implement a set of standard keywords for the config files (related to #345), then I'd be more ok with more filtering on keywords here.

Reply:
A couple things -
Which new devs do you mean?  I assume most people using using GCR don't look at register.py other than perhaps to see what services are available. People creating composite catalogs have to know about the 'catalogs' keyword.  Does anyone else?

We've already started down the road of greater control of keywords because of the special handling of the members of _PATH_DEFINING_KEYS.   `catalogs` also demands special handling because its dict may have keywords which are themselves in _PATH_DEFINING_KEYS _and are meant to be handled the same way_ as far as root dir is concerned.   If someone invents a new structure for some other config and that also happens to have a `filepath:`, `base_dir:`, etc. which is neither at the top-level or inside a `catalogs` block, there are two ways to approach it:

1. Insist that all occurrences of these keywords, no matter what the context, are treated the same; that is, always prepend root dir and don't allow absolute path in tagged versions. 
2. Only treat these keywords that way in circumstances where that is definitely the intent. Currently that means either at the top level or within a `catalogs` block.  If someone wants to invent a new block which has behavior like the `catalogs` block, they make a pull request so it can be added to _DICT_LIST

In either case we're imposing standards. I think the second approach gives developers a bit more freedom, but the first is also ok as long as everyone understands that all these keywords are behaving like reserved words, no matter what the context.